File: backend/schemas/empleado.py
"""
Schemas Pydantic para el módulo de empleados.
EmpleadoBase → EmpleadoCreate → EmpleadoUpdate → EmpleadoResponse → EmpleadoListResponse
"""
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator


def _normalizar_roles(v: object) -> object:
    """Normaliza la lista de roles: trimea, descarta vacíos y deduplica preservando orden.

    Exige al menos un rol no vacío (espejo del CHECK empleados_roles_no_vacio en DB).
    Se aplica en EmpleadoBase y EmpleadoUpdate. Mantiene el orden de carga: el primer
    elemento es el rol principal."""
    if not isinstance(v, list):
        raise ValueError("roles debe ser una lista de strings")
    limpio: List[str] = []
    for item in v:
        if not isinstance(item, str):
            raise ValueError("cada rol debe ser un string")
        texto = item.strip()
        if texto and texto not in limpio:
            limpio.append(texto)
    if not limpio:
        raise ValueError("Se requiere al menos un rol")
    return limpio


class EmpleadoBase(BaseModel):
    nombre: str
    apellido: str
    email_corporativo: str
    area_id: UUID
    roles: List[str]        # multi-valor; roles[0] es el principal (unifica cargo + rol)
    modalidad_trabajo: str  # presencial | remoto | hibrido
    tipo_contrato: str      # efectivo | plazo_fijo | contratado | pasantia
    fecha_ingreso: date
    cargo: Optional[str] = None  # DEPRECADO (se dropea en S6); lo migran S2/S5
    rol: Optional[str] = None    # DEPRECADO (se dropea en S6); lo migran S2/S5
    # Legajo ampliado (A1.1, migración 060) — todos opcionales, texto libre.
    email_personal: Optional[str] = None
    tipo_documento: Optional[str] = None
    sexo: Optional[str] = None
    telefono_alternativo: Optional[str] = None
    domicilio: Optional[str] = None
    estudios: Optional[str] = None
    ubicacion: Optional[str] = None
    turno: Optional[str] = None
    horas_contrato: Optional[int] = None
    organismo: Optional[str] = None
    gerencia: Optional[str] = None
    sector: Optional[str] = None
    seniority: Optional[str] = None
    perfil: Optional[str] = None
    categoria: Optional[str] = None
    modalidad_contratacion: Optional[str] = None
    referido: Optional[str] = None
    es_lider: bool = False

    @field_validator("roles")
    @classmethod
    def _validar_roles(cls, v: object) -> object:
        return _normalizar_roles(v)


class EmpleadoCreate(EmpleadoBase):
    empresa_id: UUID  # la empresa de pertenencia viaja en el body, no en el header
    telefono: Optional[str] = None
    fecha_nacimiento: Optional[date] = None
    dni: Optional[str] = None
    cuil: Optional[str] = None
    legajo: Optional[str] = None
    manager_id: Optional[UUID] = None  # superior inmediato (self-FK a empleados)
    dias_vacaciones_asignados: Optional[int] = None  # default 14 en DB si no se provee


class EmpleadoUpdate(BaseModel):
    nombre: Optional[str] = None
    apellido: Optional[str] = None
    email_corporativo: Optional[str] = None
    area_id: Optional[UUID] = None
    cargo: Optional[str] = None
    modalidad_trabajo: Optional[str] = None
    tipo_contrato: Optional[str] = None
    fecha_ingreso: Optional[date] = None
    telefono: Optional[str] = None
    fecha_nacimiento: Optional[date] = None
    dni: Optional[str] = None
    cuil: Optional[str] = None
    legajo: Optional[str] = None
    manager_id: Optional[UUID] = None  # superior inmediato (self-FK a empleados)
    estado: Optional[str] = None
    roles: Optional[List[str]] = None  # si se provee, reemplaza la lista completa
    rol: Optional[str] = None          # DEPRECADO (se dropea en S6)
    dias_vacaciones_asignados: Optional[int] = None
    # Legajo ampliado (A1.1, migración 060) — todos opcionales.
    email_personal: Optional[str] = None
    tipo_documento: Optional[str] = None
    sexo: Optional[str] = None
    telefono_alternativo: Optional[str] = None
    domicilio: Optional[str] = None
    estudios: Optional[str] = None
    ubicacion: Optional[str] = None
    turno: Optional[str] = None
    horas_contrato: Optional[int] = None
    organismo: Optional[str] = None
    gerencia: Optional[str] = None
    sector: Optional[str] = None
    seniority: Optional[str] = None
    perfil: Optional[str] = None
    categoria: Optional[str] = None
    modalidad_contratacion: Optional[str] = None
    referido: Optional[str] = None
    es_lider: Optional[bool] = None

    @field_validator("roles")
    @classmethod
    def _validar_roles(cls, v: object) -> object:
        return _normalizar_roles(v) if v is not None else None

    @field_validator("fecha_ingreso", "fecha_nacimiento", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: object) -> object:
        return None if v == "" else v

    @field_validator("dias_vacaciones_asignados", mode="before")
    @classmethod
    def coerce_dias(cls, v: object) -> object:
        """Convierte string numérico a int (el modal del frontend envía strings)."""
        if isinstance(v, str):
            return int(v) if v.strip() else None
        return v


class EmpleadoResponse(BaseModel):
    id: str
    nombre: str
    apellido: str
    email_corporativo: str
    empresa_id: Optional[str] = None
    empresa_nombre: Optional[str] = None
    area_id: str
    area_nombre: Optional[str] = None
    roles: List[str]                  # multi-valor; roles[0] es el principal
    modalidad_trabajo: str
    tipo_contrato: str
    fecha_ingreso: date
    telefono: Optional[str] = None
    fecha_nacimiento: Optional[date] = None
    dni: Optional[str] = None
    cuil: Optional[str] = None
    legajo: Optional[str] = None
    manager_id: Optional[str] = None      # superior inmediato (id)
    manager_nombre: Optional[str] = None  # "Apellido, Nombre" resuelto por join
    cargo: Optional[str] = None       # DEPRECADO (se dropea en S6)
    rol: Optional[str] = None         # DEPRECADO (se dropea en S6)
    estado: str
    dias_vacaciones_asignados: int = 14
    # Legajo ampliado (A1.1, migración 060) — todos opcionales.
    email_personal: Optional[str] = None
    tipo_documento: Optional[str] = None
    sexo: Optional[str] = None
    telefono_alternativo: Optional[str] = None
    domicilio: Optional[str] = None
    estudios: Optional[str] = None
    ubicacion: Optional[str] = None
    turno: Optional[str] = None
    horas_contrato: Optional[int] = None
    organismo: Optional[str] = None
    gerencia: Optional[str] = None
    sector: Optional[str] = None
    seniority: Optional[str] = None
    perfil: Optional[str] = None
    categoria: Optional[str] = None
    modalidad_contratacion: Optional[str] = None
    referido: Optional[str] = None
    es_lider: bool = False
    created_at: datetime


class EmpleadoListResponse(BaseModel):
    items: List[EmpleadoResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class EmpleadoSeleccionable(BaseModel):
    """Proyección liviana de un empleado para poblar selects (ej. superior inmediato)."""
    id: str
    nombre: str
    apellido: str

File: frontend/types/empleado.ts
export interface Empleado {
  id: string
  nombre: string
  apellido: string
  email_corporativo: string
  empresa_id: string | null
  empresa_nombre: string | null
  area_id: string
  area_nombre: string | null
  roles: string[]
  cargo?: string | null // DEPRECADO (se quita en S6); usar roles
  modalidad_trabajo: "presencial" | "remoto" | "hibrido"
  tipo_contrato: string // texto libre (migración 065); el CSV de nómina trae valores abiertos
  fecha_ingreso: string
  telefono: string | null
  fecha_nacimiento: string | null
  dni: string | null
  cuil: string | null
  legajo: string | null
  manager_id: string | null // superior inmediato (id)
  manager_nombre: string | null // "Apellido, Nombre" resuelto por el backend
  estado: "activo" | "baja" | "licencia"
  dias_vacaciones_asignados: number
  // Legajo ampliado (A1)
  email_personal: string | null
  tipo_documento: string | null
  sexo: string | null
  telefono_alternativo: string | null
  domicilio: string | null
  estudios: string | null
  ubicacion: string | null
  turno: string | null
  horas_contrato: number | null
  organismo: string | null
  gerencia: string | null
  sector: string | null
  seniority: string | null
  perfil: string | null
  categoria: string | null
  modalidad_contratacion: string | null
  referido: string | null
  es_lider: boolean
  created_at: string
}

export interface EmpleadoListResponse {
  items: Empleado[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface EmpleadoCreate {
  empresa_id: string
  nombre: string
  apellido: string
  email_corporativo: string
  area_id: string
  roles: string[]
  modalidad_trabajo: string
  tipo_contrato: string
  fecha_ingreso: string
  telefono?: string
  fecha_nacimiento?: string
  dni?: string
  cuil?: string
  legajo?: string
  manager_id?: string // superior inmediato (id)
  cargo?: string // DEPRECADO (se quita en S6); el form ya no lo manda
  rol?: string // DEPRECADO (se quita en S6)
  dias_vacaciones_asignados?: number
  // Legajo ampliado (A1) — todos opcionales
  email_personal?: string
  tipo_documento?: string
  sexo?: string
  telefono_alternativo?: string
  domicilio?: string
  estudios?: string
  ubicacion?: string
  turno?: string
  horas_contrato?: number
  organismo?: string
  gerencia?: string
  sector?: string
  seniority?: string
  perfil?: string
  categoria?: string
  modalidad_contratacion?: string
  referido?: string
  es_lider?: boolean
}

export type EmpleadoUpdate = Partial<EmpleadoCreate> & { estado?: string }

/** Proyección liviana de empleado para poblar selects (ej. superior inmediato). */
export interface EmpleadoSeleccionable {
  id: string
  nombre: string
  apellido: string
}
